lib/mauth/faraday.rb

```diff
@@ -68,6 +75,14 @@ def x_mws_time
       def x_mws_authentication
         @response_env[:response_headers]['x-mws-authentication']
       end
+
+      def mcc_time
```

Comment:
where is this used?

Reply:
it's used in the v2 time_validation and signature validation in `client.rb`.

Reply:
I know I'm saying the same thing again but as is now we are adding v2 only methods to this class, which force us to add v2 methods or even mixed v1-v2 methods to the Signed module. And I guess the same happens with authorization. I think we could have different classes `ResponseV2` for instance instead

Reply:
@mdsol/team-16 I talked about creating `ResponseV2` and `RequestV2` classes and only including the relevant methods in those class with @jcarres-mdsol. I think it would improve readability and maintainability in some files. One problem is that during the conversion process while we're signing with both V1 and V2 we will have to instantiate two request objects every time we sign. We'll also have to move the logic of deciding which protocol to sign with into both the Faraday and rack middlewares. Thoughts?

Reply:
@cmcinnes-mdsol @jcarres-mdsol In theory I like the idea of creating `ResponseV2` and `RequestV2` (and also `ResponseV1` and `RequestV1` classes and having common code elsewhere -- I also like the idea of moving the `Signer` methods out of the `Client` so we can just do something like `request.authentic?`, `response.sign!`, `request.sign!` or `response.authentic?` in the middlewares and have those objects delegate to implementations of the proper version of the protocol).  In practice, I am concerned about the risks of changing the code too much and spending additional time on this now.  While the existing code may not be as clear as we'd like it and the files too big, it has been running in production for several years and I worry that a big refactor will introduce bugs that are hard to find.  Or do you think the additions that have been made already are already very disruptive to the V1 code path?

As for creating two objects with every signature, I think we would need a `Benchmark` to know if that causes much slower performance.  For memory usage due to multiple objects,  you could try something like tje `memory_profiler` gem.

If we do go ahead with this refactor, I would prefer to see it in a separate PR, with the hope that the specs that were added here would not have to change too much to just pass after the refactor. 🤞 Having a separate PR also might facilitate doing the benchmarking and memory profiling comparison.

Reply:
FWIW I am pretty much in agreement with @ejennings-mdsol.. been trying to wrap my head around this PR for days without much success.

There is much that I hate about this gem, with the client file in particular being an abomination: not actually a client but rather just a hideous way of gluing together signing/authenticating with config.

I am not a huge fan of the OO stuff in here either, so I'd be ok with just slapping all of the V2 functionality onto the same `Request` object for now.  In a (later / other story) reorganization I'd look to get rid of the object entirely and just have two modules for signing and authenticating that expose one function each, then let the respective middlewares worry about how/where to attach those results (since that is their business anyway).

I think the various requirements of the story, in particular the "only" flags, may be causing this PR to be so muddled.  Perhaps we could split in PRs like (1) add the basic v2 stuff and always sign with / accept both, (2) add the flags, (3) add documentation?